test/intfuncs.jl

```diff
@@ -191,6 +191,13 @@ end
 
         @test lcm(T[2, 4, 6]) ⟷ T(12)
     end
+
+    # Issue #55379
+    @test lcm([1//2; 1//2]) === lcm([1//2, 1//2]) === lcm(1//2, 1//2) === 1//2
```

Comment:
IMO, this is wrong. Rational numbers are a field, and as such to not have an LCM

Reply:
xref #56166

Reply:
I think it is perfectly sensible to define lcm on rationals, provided we choose a definition for "multiple" that is based on integers. A definition that is mathematically sensible and in line with current behavior (except for #55379) is

For a signed ring `R`, `m` in `R` is an _lcm_ of `a_1, a_2, ... a_n in R` iff `m` is positive and is a multiple of `a_1`, `a_2`, ... and `a_n` and for all other `m'` in `R` that are multiples of `a_1`, `a_2`, ... and `a_n`, `m'` is a multiple of `m`.

 And we define "multiple" as

`a` is a _multiple_ of `b` iff there exists an integer `n` such that `b * n = a` where `*` refers to multiplication in the ring of integers, not in the ring of integers mod 2^64.

---

While there is certainly a case to be made that `lcm` should throw on rationals, now is not the time to make that case. That time was #33910. `lcm` has supported rationals under the definition above since 1.4.

---


I am not aware of any definitions of `lcm` over the integers or the rationals that would result in returning different answers; though some definitions should throw errors:

Using "`m <= m'`" instead of "`m'` is a multiple of `m`" produces the same results on rationals and integers.

Defining "multiple" based on the existence of a ring element that takes `b` to `a` instead of based on an integer that takes `b` to `a` is equivalent for the ring of integers. For the rationals, every positive rational is the `lcm` of every set of rationals which doesn't seem like a particularly useful definition. Folks sometimes (e.g. https://github.com/JuliaLang/julia/issues/27039#issuecomment-387634155, https://github.com/JuliaLang/julia/issues/56166#issuecomment-2430411620) say that `lcm(a,b) == 1` for a field which, if we are defining multiple as "all field elements are multiples of all other field elements", is not wrong but we could just as correctly say that `lcm(a,b) == 17` for all field elements `a` and `b` under that definition which means the right answer for a programming language is an error, not the number 1 nor the number 17.

Reply:
Ah yes, the [FTA definition](https://en.wikipedia.org/wiki/Least_common_multiple#Fundamental_theorem_of_arithmetic) is nice. AFAICT these two definitions agree in all cases. 

For the empty case over the rationals the definition I listed indicates there is no LCM. Using the FTA definition extended to rationals requires computing `max` over an empty set of integers which also does not exist.

Reply:
I think it would be `prod` over an empty set, right? which does exist and `== 1`

Reply:
Interesting, I was thinking of representing a rational number (e.g. `5//7`) as an infinite sequence of prime powers (e.g. `0,0,1,-1,0,0,0,...`) in which case to compute `lcm`, one takes element wise max over those sequences and then converts back to a rational. In the empty case it would be a max over an empty set. 

If, instead, one represents rationals as finite sets of prime-power pairs (e.g. `5^1 * 7^(-1)`), then to compute the `lcm(x1, x2, ...)` one would take the union of the primes represented in each `x`, for each prime compute the max power, and take the product of those primes. In the empty case, the union of the primes represented in each `x` is the empty set and one would then take get `prod` over the empty set and get `1`.

I prefer using infinite sequences because I think it's simpler and more elegant but that is entirely subjective and it is definitely mathematically sensible to use the FTA definition, decompose to finite sets, and get `lcm(Rational[]) == 1//1`.

Given that we already have

```julia-repl
julia> lcm(Rational[])
1//1
```

and that `1//1` is a reasonable answer, this PR's change to make that invocation throw is breaking.

Reply:
as counterpoint to my interpretation, it would lead to `gcd([]) = 1` as well, which is normatively worse than `gcd([]) = 0`

to be honest though, I think all these choices are basically just convention and I don't think it matters _that_ much that there is some underlying purity determining the convention as long as what we end up with makes sense.

I think the current state of this PR makes good choices for each convention, and afaict it looks like the court of public opinion --- which seems to arise from the recursive definitions --- agrees https://math.stackexchange.com/questions/1755266/gcd-of-an-empty-set https://www.reddit.com/r/learnmath/comments/v9vmfm/whats_the_empty_lcm/

Reply:
@adienes, please take a look at https://github.com/JuliaLang/julia/issues/56166

> all these choices are basically just convention and I don't think it matters that much that there is some underlying purity determining the convention as long as what we end up with makes sense

The only way to end up with something that makes sense is to respect the math ("underlying purity").

Reply:
respectfully, I don't think "the math" makes any particularly consistent demands about what to do in these edge cases. and any choice made is essentially just convention (which can and **_does_** vary among authors)

Reply:
As described in the linked issue, there indeed are multiple possible choices here. That doesn't imply that *anything goes*, though.

Reply:
I think a more charitable interpretation of my comments (rather than "anything" goe") is "a lot of things could go, and I think the choices made here are good ones"

but anyway I guess let's just leave to to triage. I don't typically attend triage but if the discussion comes up, I would upvote the implementation in this PR.